### PR TITLE
Device Disk

### PR DESCRIPTION
**Pull request by tomponline**
Migration of disk device into the new device interface.

Adds following behaviour changes:

- Host-side mounts for custom volumes are now unmounted after device start (like rootfs is).
- The `optional` property has been deprecated and replaced with `required` property (defaulting to true) to be consistent with other device types. The `optional` property still works but it has been removed from docs.
- Adds new `CGroupBlkioWeightController` check for `limits.disk.priority`.
- If disk limits cannot be applied due to lack of host cgroup controllers then container wont start.

Part of #5819

- [x] RootFS
- [x] Custom Volumes
- [x] I/O limits
- [x] Live update of limits
- [x] Size quota limits

### CODE REVIEW

**Comment by tomponline** [comment]
@stgraber in the interests of time, you may want to take a look at this now and then the disk io limits stuff can follow on as a separate PR, as this is getting quite big as it is.

File: lxd/storage.go [hunk 1]
```diff
@@ -873,3 +882,41 @@ func storagePoolDriversCacheUpdate(cluster *db.Cluster) {
...
+		return err
+	}
+
+	return nil
+}
+
+// storageVolumeUmount unmounts a storage volume on a pool.
+func storageVolumeUmount(state *state.State, poolName string, volumeName string, volumeType int) error {
+	// tomp TODO check about the "default" hardcoded project here.
+	s, err := storagePoolVolumeInit(state, "default", poolName, volumeName, volumeType)
```
Comment (tomponline) [inline lxd/storage.go]: @stgraber I noticed that the existing code I ported over has the project hardcoded as "default", is this expected here?

File: lxd/container.go [hunk 2]
```diff
@@ -214,118 +172,19 @@ func containerValidConfig(sysOS *sys.OS, config map[string]string, profile bool,
...
-		if !shared.StringInSlice(m["type"], []string{"disk", "none"}) {
-			return fmt.Errorf("Invalid device type for device '%s'", name)
-		}
-
-		for k := range m {
-			if !containerValidDeviceConfigKey(m["type"], k) {
-				return fmt.Errorf("Invalid device configuration key for %s: %s", m["type"], k)
-			}
+// containerValidDevices validate container device configs.
+func containerValidDevices(state *state.State, cluster *db.Cluster, instance device.InstanceIdentifier, devices deviceConfig.Devices, expanded bool) error {
```
Comment (tomponline) [inline lxd/container.go]: @stgraber I've removed the `profile` boolean argument from this function, as each device can use the presence of an empty instance name to indicate profile validation mode (if needed).

However I am unclear on the role of the `expanded` boolean in this function, and how it relates to disk validation. As it is, the new code doesn't take this into account and proceeds with normal validation irrespective of this boolean value, so I would like to remove it ideally, but I am unclear what it is for.

File: lxd/storage.go (hunk 1 above)
Comment (stgraber) [reply lxd/storage.go]: For custom storage volumes, yes, it is. We don't currently have those namespaced per-project.

File: lxd/container.go (hunk 2 above)
Comment (stgraber) [reply lxd/container.go]: I believe we sorted that part out on IRC, right?

**Comment by stgraber** [comment]
Looks good to me, obviously we need to address the deviceConfig branch first, then this one will shrink quite a bit, but looking at those commits which aren't in the deviceConfig branch, things look good.

File: lxd/container.go (hunk 2 above)
Comment (tomponline) [reply lxd/container.go]: @stgraber yes thanks, I've implemented that now. I've still kept the `expanded` boolean on this function and use that to indicate how to populate a temporary `containerLXC` struct such that the population of `expandedDevices` indicates an expanded validation.

**Pull request merged**